### PR TITLE
[RLlib] Fix update model view requirements from init state for bare-metal policies with custom view-reqs.

### LINKED ISSUE #17866: User-defined `view_requirements` get overwritten when inheriting directly from `Policy`.
### What is the problem?

When writing a policy that directly inherits from `Policy` and using the *View Trajectory API* the super class `Policy` calls `_update_view_requirements_from_init_state()` which results in user-defined `view_requirements` for the `state_in_0` column getting overwritten and consequently ignored. 

*Ray version and other system information (Python version, TensorFlow version, OS):*

OS: Fedora 34
Python version: 3.9.6
Ray version: 2.0.0.dev0

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
import numpy as np
import gym
import ray
from gym.spaces import Box
from ray.rllib.policy import Policy
from ray.rllib.utils.annotations import override
from ray.rllib.utils.typing import ModelWeights
from ray.rllib.policy.view_requirement import ViewRequirement
from ray.rllib.agents.trainer_template import build_trainer

class MyPolicy(Policy):
    def __init__(self, observation_space, action_space, model_config, *args, **kwargs):
        super(MyPolicy, self).__init__(observation_space, action_space, model_config, *args, **kwargs)
        self.observation_space = observation_space
        self.action_space = action_space
        self.model_config = model_config or {}
        space = Box(low=-np.inf, high=np.inf, shape=(10,), dtype=np.float64)
        self.view_requirements['state_in_0'] = \
            ViewRequirement('state_out_0',
                            shift="-{}:-1".format(2),
                            used_for_training=False,
                            used_for_compute_actions=True,
                            batch_repeat_value=1)
        self.view_requirements['state_out_0'] = \
            ViewRequirement(
                    space=space, 
                    used_for_training=False,
                    used_for_compute_actions=True,
                    batch_repeat_value=1)

    def get_initial_state(self):
        return [
            np.zeros((10,), dtype=np.float32)
        ]

    def compute_actions(self,
                        obs_batch=None,
                        state_batches=None,
                        prev_action_batch=None,
                        prev_reward_batch=None,
                        info_batch=None,
                        episodes=None,
                        **kwargs):
        
        actions = [np.power(np.random.random(10), state_batch[0]) for state_batch in state_batches[0]]
        new_state_batches = [action + state_batch[1] for action, state_batch in zip(actions, state_batches[0])]
        return actions, new_state_batches, {}
    
    def learn_on_batch(self, samples):
        return

    @override(Policy)
    def get_weights(self) -> ModelWeights:
        """No weights to save."""
        return {}

    @override(Policy)
    def set_weights(self, weights: ModelWeights) -> None:
        """No weights to set."""
        pass

MyTrainer = build_trainer(
    name="MyPolicy",
    default_policy=MyPolicy)

class MyEnv(gym.Env):  
    def __init__(self, env_config=None):
        super(MyEnv, self).__init__()
        self.config = env_config or {}
        self.observation_space = Box(low=-1.0, high=1.0, shape=(10,), dtype=np.float32)
        self.action_space = Box(low=-np.Inf, high=np.Inf, shape=(10,), dtype=np.float32)

    def reset(self):
        self.timestep = 0
        return self.observation_space.sample()

    def step(self, action):
        self.timestep += 1 
        new_obs = self.observation_space.sample()        
        reward = np.dot(new_obs, action)
        if self.timestep % 50 == 0:
            print('Reward: {}'.format(reward))
        if reward > 1.0 or reward < -1.0:
            done = True
        else: 
            if self.timestep > 1000:
                done= True
            else:
                done = False
        return new_obs, reward, done, {}

config = {
    "env": MyEnv,
    "model": {
        "max_seq_len": 1, # Necessary to get the whole trajectory of 'state_in_0' in the sample batch
    },
    "num_workers": 1, 
    "framework": None,  # NOTE: Does this have consequences? I use it for not loading tensorflow/pytorch
    "log_level": "DEBUG",
    "create_env_on_driver": True, 
    "batch_mode": "complete_episodes" ,
    "rollout_fragment_length": 1000,   # Each episode takes exactly 1000 steps. 
    "train_batch_size": 10000,  # We want to have 50 episodes run.
    "evaluation_num_episodes": 0, # No evaluation for deterministic policy. 
    "normalize_actions": False, # Actions are explicit and no logits. 
}

ray.init(ignore_reinit_error=True, local_mode=True)
my_trainer = MyTrainer(config=config)
results = my_trainer.train()
```
If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by simonsays1980**
…the 'shift' in view_requirements from a user-defined policy that inherits directly from Policy.

## Why are these changes needed?
When writing a policy that directly inherits from `Policy` and using the *View Trajectory API* the super class `Policy` calls `_update_view_requirements_from_init_state()` which results in user-defined `view_requirements` for the `state_in_0` column getting overwritten and consequently ignored. 

The fix only sets the view_requirements for the `state` if the user has not given her own. 

## Related issue number
Closes [17866](https://github.com/ray-project/ray/issues/17866)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
This is great @simonsays1980 , thanks for this PR! :)
If you have a few minutes, maybe we could add another PR with a short example script that includes:
- very simple custom policy that defines its own view requirements.
- in the policy's compute_actions_from_input_dict method, place an assertion that the incoming shape of the input is correct (according to the view requirements).
- tie this policy into a custom trainer and train for a few iterations.

**Review by sven1977 — APPROVED** [review]
LGTM! :)

File: rllib/policy/policy.py
Comment (sven1977) [inline rllib/policy/policy.py:937]:
```suggestion
                # Only override if user has not already provided
                # custom view-requirements for state_in_n.
                if "state_in_{}".format(i) not in vr:
```

File: rllib/policy/policy.py
Comment (sven1977) [inline rllib/policy/policy.py:945]:
```suggestion
                # Only override if user has not already provided
                # custom view-requirements for state_out_n.
                if "state_out_{}".format(i) not in vr:
```

**Comment by sven1977** [comment]
Tests look actually good. Just waiting for the re-run before we can merge this.

**Comment by simonsays1980** [comment]
> This is great @simonsays1980 , thanks for this PR! :)
> If you have a few minutes, maybe we could add another PR with a short example script that includes:
> 
> * very simple custom policy that defines its own view requirements.
> * in the policy's compute_actions_from_input_dict method, place an assertion that the incoming shape of the input is correct (according to the view requirements).
> * tie this policy into a custom trainer and train for a few iterations.

@sven1977 you mean as an *example* for others inheriting their class directly from the `Policy`?

**Pull request merged**
